contracts/src/instances/savings-usds-l2/SavingsUSDSL2Base.sol

```diff
@@ -0,0 +1,164 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity 0.8.24;
+
+import { ERC20 } from "openzeppelin/token/ERC20/ERC20.sol";
+import { SafeERC20 } from "openzeppelin/token/ERC20/utils/SafeERC20.sol";
+import { IPSM } from "../../interfaces/IPSM.sol";
+import { IHyperdrive } from "../../interfaces/IHyperdrive.sol";
+import { HyperdriveBase } from "../../internal/HyperdriveBase.sol";
+import { SavingsUSDSL2Conversions } from "../../instances/savings-usds-l2/SavingsUSDSL2Conversions.sol";
+
+/// @author DELV
+/// @title SavingsUSDSL2Base
+/// @notice The base contract for the SavingsUSDSL2 Hyperdrive implementation.
+/// @dev This Hyperdrive implementation is designed to work with standard
+///      SavingsUSDSL2 vaults. Non-standard implementations may not work correctly
+///      and should be carefully checked.
+/// @custom:disclaimer The language used in this code is for coding convenience
+///                    only, and is not intended to, and does not, have any
+///                    particular legal or regulatory significance.
+abstract contract SavingsUSDSL2Base is HyperdriveBase {
+    using SafeERC20 for ERC20;
+
+    IPSM internal immutable _PSM;
+
+    /// @notice instantiates the SavingsUSDSL2Base contract.
+    /// @param __PSM The PSM contract.
+    constructor(IPSM __PSM) {
+        // Initialize the PSM immutable.
+        _PSM = __PSM;
+    }
+
+    /// Yield Source ///
+
+    /// @dev Accepts a deposit from the user in base.
+    /// @param _baseAmount The base amount to deposit.
+    /// @return The shares that were minted in the deposit.
+    /// @return The amount of ETH to refund. Since this yield source isn't
+    ///         payable, this is always zero.
+    function _depositWithBase(
+        uint256 _baseAmount,
+        bytes calldata // unused
+    ) internal override returns (uint256, uint256) {
+        // Take custody of the deposit in base.
+        ERC20(address(_baseToken)).safeTransferFrom(
+            msg.sender,
+            address(this),
+            _baseAmount
+        );
+
+        // Deposit the base into the yield source.
+        //
+        // NOTE: We increase the required approval amount by 1 wei so that
+        // the vault ends with an approval of 1 wei. This makes future
+        // approvals cheaper by keeping the storage slot warm.
+        ERC20(address(_baseToken)).forceApprove(address(_PSM), _baseAmount + 1);
+
+        // Depositing amounts to swapping USDS for SUSDS in the PSM.
+        uint256 sharesMinted = _PSM.swapExactIn(
+            address(_baseToken),
+            address(_vaultSharesToken),
+            _baseAmount,
+            _convertToShares(_baseAmount),
+            address(this),
+            0
```

Comment:
Might be good to reach out to Spark for a referral code. Charles would be a good person to ask about this. It doesn't need to block the PR, but I don't want us to forget about this